File: Iteration2/SourceCode/src/main/java/Student.java
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;

import java.io.*;
import java.nio.channels.FileChannel;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Map;
import java.util.TreeMap;

public class Student {
    private Identity id;
    private int academicYear=0;
    private Advisor advisor;
    private Transcript transcript;
    private Schedule schedule;
    private int academicSemester=0;


    public Student(Identity id, int academicYear,Advisor advisor,Transcript transcript,Schedule schedule) {
        this.id = id;
        this.transcript = transcript;
        this.academicYear = academicYear;
        this.advisor=advisor;
        this.setSchedule(schedule);
    }

    public Identity getId() {
        return id;
    }

    public void setId(Identity id) {
        this.id = id;
    }

    public Transcript getTranscript() {
        return transcript;
    }




    public void setTranscript(Transcript transcript) {
        this.transcript = transcript;
    }

    public int getAcademicYear() {
        return academicYear;
    }

    public void setAcademicYear(int academicYear) {
        this.academicYear = academicYear;
    }

    public int getAcademicSemester() {
        return academicSemester;
    }

    public void setAcademicSemester(int academicSemester) {
        this.academicSemester = academicSemester;
    }




    public void startRegistration(ArrayList<Semester> allSemesterList,SemesterStatistic semesterStatistic) {
        ArrayList<Course> failedCourses = this.getTranscript().getFailedCourses();
        ArrayList<Course> completedCourses = this.getTranscript().getCompletedCourses();
        ArrayList<String> preqCourses = new ArrayList<>();
        // Registering to failedCourses first.
        for (Course course : failedCourses) {
            ArrayList<CourseSection> sectionsOfCourse = course.getCourseSection();
            int CourseSectionCount = sectionsOfCourse.size();
            int sectionNumber = (int) (Math.random() * CourseSectionCount)+1;

        } // Registering to currentSemesterCourses

        for (Semester semester : allSemesterList){
            ArrayList<Course> currentSemesterCourses = semester.getCourseList();
            for (Course course : currentSemesterCourses) {
                if(course.getSemester()<=this.getAcademicSemester()+1 &&  !completedCourses.contains(course)) {
                    ArrayList<CourseSection> sectionsOfCourse = course.getCourseSection();
                    int CourseSectionCount = sectionsOfCourse.size();
                    int sectionNumber = (int) (Math.random() * CourseSectionCount) + 1; // this val is to register randomly

                    ArrayList<Course> prerequisites = course.getPrerequisite();
                    boolean hasPrerequisite = false;
                    for (Course pCourse : prerequisites) {
                        if (completedCourses.contains(pCourse) ) {
                            hasPrerequisite = true;
                        } else {
                            hasPrerequisite = false;
                            break;
                        }
                    }
                    if(prerequisites.size()==0) {
                        hasPrerequisite=true;
                    }

                    if (hasPrerequisite ) {
                        sectionsOfCourse.get(sectionNumber - 1).requestToRegister(this);

                    } else {
                        preqCourses.add("The system didn't allow "+sectionsOfCourse.get(sectionNumber-1).getCourse().get(0).getCode()+"  because student failed prereq.");
                        CourseSectionRegister register = new CourseSectionRegister(sectionsOfCourse.get(sectionNumber-1),"PREREQUISITE",this);
                        semesterStatistic.addtoDeniedCourseSectionRegisterList(register);
                    }
                }
            }

        }
        this.advisor.checkStudentSchedule(this,semesterStatistic,preqCourses);
    }
    public void writeToTheJSON(Map<String,String> courseMap){
        String fileName = getId().getID()+".json";JSONObject jsonObject = new JSONObject();
        JSONArray courses = new JSONArray();
        for (Map.Entry<String,String> entry : courseMap.entrySet()){
            courses.add("Course: " + entry.getKey() +", Grade:"+ entry.getValue());
        }

        //logs.add("Course: " + name +", Grade:"+ grade );
        jsonObject.put("Semester " + (this.academicSemester+1) + ": ",courses );
        if(this.academicSemester+1==1){
            try {
                FileWriter file = new FileWriter(fileName,true);
                file.write("[" +jsonObject.toJSONString()+"\n,");
                file.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        else{

            try {
                FileWriter file = new FileWriter(fileName,true);
                file.write(jsonObject.toJSONString()+"\n,");
                file.close();
            } catch (IOException e) {
                e.printStackTrace();
            }}
    }
    public void writeLogs(ArrayList<String> logs){
        String fileName = getId().getID()+".json";JSONObject jsonObject = new JSONObject();
        JSONArray logOfCourses = new JSONArray();
        logs.forEach((n) -> logOfCourses.add(n));
        jsonObject.put("Logs: ",logOfCourses);
        try {
            FileWriter file = new FileWriter(fileName,true);
            file.write(jsonObject.toJSONString()+"\n,");
            file.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
    public void writeGPA(String logs){
        String fileName = getId().getID()+".json";JSONObject jsonObject = new JSONObject();
        JSONArray logOfCourses = new JSONArray();
        logOfCourses.add(logs);
        jsonObject.put("GPA: ",logOfCourses);
        try {
            FileWriter file = new FileWriter(fileName,true);
            file.write(jsonObject.toJSONString()+"\n,");
            file.close();
        } catch (IOException e) {
            e.printStackTrace();
        }

    }
    public void endOfFile(){
        String fileName = getId().getID()+".json";JSONObject jsonObject = new JSONObject();
        try {
            File file = new File(fileName);
            FileChannel fileChannel = new FileOutputStream(file, true).getChannel();
            fileChannel.truncate(fileChannel.size() - 1); //Removes last character
            fileChannel.close();
        }
        catch (FileNotFoundException ex) {

        }
        catch (IOException ex) {

        }
        try {
            FileWriter file = new FileWriter(fileName,true);
            file.write("]");
            file.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void finishSemester(){
        ArrayList<CourseSectionRegister> registerList = this.schedule.getCourseSectionRegisterList();
        ArrayList<Course> courses= new ArrayList<Course>();
        int ntecount=0;
        int ftecount=0;
        int tecount=0;
        for(int i=0;i<getTranscript().getGradeBook().size();i++){
            if(!(getTranscript().getGradeBook().get(i).getLetterGrade().equals("FF")||getTranscript().getGradeBook().get(i).getLetterGrade().equals("FD"))) {

                if (getTranscript().getGradeBook().get(i).getCourse().getType().equals("NTE")) {
                    ntecount++;
                } else if (getTranscript().getGradeBook().get(i).getCourse().getType().equals("TE")) {
                    tecount++;
                }
            }
        }
        for(CourseSectionRegister register: registerList){
            CourseSection sectionOfReg= register.getCourseSection();
            Course courseOfSection = sectionOfReg.getCourse().get(0);
            if(sectionOfReg.getCourse().get(0).getType().equals("FTE")){
                 courseOfSection = sectionOfReg.getCourse().get(ftecount);
            }
            else if(sectionOfReg.getCourse().get(0).getType().equals("TE")){
                 courseOfSection = sectionOfReg.getCourse().get(tecount);
                 tecount++;
            }
            else if(sectionOfReg.getCourse().get(0).getType().equals("NTE")){
                 courseOfSection = sectionOfReg.getCourse().get(ntecount);
            }

            courses.add(courseOfSection);
        }

        Map<String,String> courseMap = new TreeMap<String,String>();
        String gpaJSON;
        gpaJSON = "GPA: "+ this.transcript.addCoursesToTranscript(courses,courseMap)+",cGPA: "+getTranscript().calculateTotalGPA();

        writeToTheJSON(courseMap);
        writeGPA(gpaJSON);
        courseMap.clear();
        this.schedule.clearSchedule();
        this.incrSemester();
    }


    public Schedule getSchedule() {
        return schedule;
    }

    public void setSchedule(Schedule schedule) {
        this.schedule = schedule;
    }

    public void incrSemester(){
        this.setAcademicSemester((this.getAcademicSemester()+1));
    }
}

File: Iteration3/SourceCode/Student.py
import random
import os
import time

from CourseSectionRegister import CourseSectionRegister
from Advisor import Advisor
from Lecturer import Lecturer
from Transcript import Transcript
import json
# encoding:utf-8


class Student(object):


    def __init__(self, id, academicYear, advisor, transcript, schedule):
        advisor = Advisor(None,None,None)
        self.__academicSemester =0
        self.__id = id
        self.__transcript = transcript
        self.__academicYear = academicYear
        self.__advisor=advisor
        self.setSchedule(schedule)

    def getId(self):
        return self.__id

    def setId(self, id):
        self.__id = id

    def getTranscript(self):
        return self.__transcript

    def setTranscript(self, transcript):
        self.__transcript = transcript

    def getAcademicYear(self):
        return self.__academicYear

    def setAcademicYear(self, academicYear):
        self.__academicYear = academicYear

    def getAcademicSemester(self):
        return self.__academicSemester

    def setAcademicSemester(self, academicSemester):
        self.__academicSemester = academicSemester

    def startRegistration(self, allSemesterList, semesterStatistic):
        failedCourses=[]
        failedCourses = self.getTranscript().getFailedCourses()
        completedCourses = self.getTranscript().getCompletedCourses()
        preqCourses = []
        for course in failedCourses:
            sectionsOfCourse = course.getCourseSection()
            CourseSectionCount = len(sectionsOfCourse)
            #sectionNumber = 545454 #random

        for semester in allSemesterList:
            currentSemesterCourses = semester.getCourseList()
            for course2 in currentSemesterCourses:
                if course2.getSemester()<=self.getAcademicSemester()+1 and (course2 not in completedCourses):
                    sectionsOfCourse = course2.getCourseSection()
                    CourseSectionCount = len(sectionsOfCourse)
                    sectionNumber = int(random.randint(0,CourseSectionCount-1))+1 # random
                    prerequisites = course2.getPrerequisite()
                    hasPrerequisite = False
                    for pCourse in prerequisites:
                        if pCourse in completedCourses:
                            hasPrerequisite = True
                        else:
                            hasPrerequisite = False
                            break

                    if len(prerequisites)==0:
                        hasPrerequisite=True

                    if hasPrerequisite:
                        sectionsOfCourse[sectionNumber-1].requestToRegister(self)

                    else:
                        preqCourses.append("The system didn't allow "+sectionsOfCourse[sectionNumber-1].getCourse()[0].getCode()+" because student failed prereq.")
                        register = CourseSectionRegister(sectionsOfCourse[sectionNumber-1], "PREREQUISITE", self)
                        semesterStatistic.addtoDeniedCourseSectionRegisterList(register)

        self.__advisor.checkStudentSchedule(self, semesterStatistic,preqCourses)


    def writeToTheJSON(self, courseMap):
        name = str(self.getId().getID())+'.json'

        if(bool(courseMap)==False):
            courseMap={"": ""}


        if (os.path.isfile(name))==False:
            ss={}
            with open(name, "w", encoding='utf-8') as file:
                ss["Semester "+str(self.getAcademicSemester()+1)+":"]=courseMap
                file.seek(0)

                json.dump(ss, file,indent=4)

                return
        if (os.path.isfile(name)) == True:
            with open(name, "r+",encoding='utf-8') as file:
                data = json.load(file)
                data["Semester " + str(self.getAcademicSemester()+1)+":"]=courseMap
                file.seek(0)
                json.dump(data, file,indent=4)







    def writeLogs(self, logs):
        name = str(self.getId().getID()) + '.json'

        with open(name, "r+", encoding='utf-8') as file:
            data = json.load(file)
            data["Semester " + str(self.getAcademicSemester() + 1) + " logs:"] = logs
            file.seek(0)
            json.dump(data, file, indent=4)

        return

    def writeGPA(self, logs):
        name = str(self.getId().getID())+'.json'


        with open(name, "r+", encoding='utf-8') as file:
            data = json.load(file)
            data["Semester " + str(self.getAcademicSemester()+1)+" GPA and cGPA:"] = logs
            file.seek(0)
            json.dump(data, file, indent=4)

        return



    def finishSemester(self):
        registerList = self.schedule.getCourseSectionRegisterList()
        courses = []
        ntecount=0
        ftecount=0
        tecount=0
        for i in range(0,len(self.getTranscript().getGradeBook())):
            if not (self.getTranscript().getGradeBook()[i].getLetterGrade()=="FF" or self.getTranscript().getGradeBook()[i].getLetterGrade()=="FD"):
                if self.getTranscript().getGradeBook()[i].getCourse().getType()=="NTE":
                    ntecount +=1
                elif (self.getTranscript().getGradeBook()[i].getCourse().getType()=="TE"):
                    tecount +=1

        for register in registerList:
            sectionOfReg = register.getCourseSection()
            courseOfSection = sectionOfReg.getCourse()[0]
            if (sectionOfReg.getCourse()[0].getType()=="FTE"):
                courseOfSection=sectionOfReg.getCourse()[ftecount]
            elif (sectionOfReg.getCourse()[0].getType()=="NTE"):
                courseOfSection=sectionOfReg.getCourse()[ntecount]

            courses.append(courseOfSection)


        courseMap = {}
        gpaJSON = "GPA: " + self.__transcript.addCoursesToTranscript(courses, courseMap) + ",cGPA: " + self.getTranscript().calculateTotalGPA();

        #courseDict['GPA and cGPA']= gpaJSON
        self.writeToTheJSON(courseMap);
        self.writeGPA(gpaJSON);
        courseMap.clear();
        self.schedule.clearSchedule();
        self.incrSemester();


    def getSchedule(self):
        return self.schedule

    def setSchedule(self, schedule):
        self.schedule = schedule

    def incrSemester(self):
        self.setAcademicSemester(self.getAcademicSemester()+1)
